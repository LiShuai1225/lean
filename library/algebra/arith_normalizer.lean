definition mk_monomial [irreducible] {A : Type} [has_mul A] (c pp : A) : A := c * pp
notation `[` x `,` y `]` := mk_monomial x y


check Î» x:nat, [1, x]
