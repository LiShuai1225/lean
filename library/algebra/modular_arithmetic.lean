/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Daniel Selsam
-/

structure has_modular_arith [class] (A : Type) [has_mod A] (n : ℕ) :=
(eq_modulo : ∀ (a : A), a = (a % n))

-- TODO(dhs): put into actual bit-vector development

constant (nat_pow : ℕ → ℕ → ℕ)
constant (bv : Π (n : ℕ), Type.{1})
constant (bmod : Π {n:ℕ}, bv n → ℕ → bv n)
constant (b_eq_modulo : ∀ (n:ℕ) (v : bv n), v = bmod v (nat_pow 2 n))

definition bv_has_mod [instance] (n : ℕ) : has_mod (bv n) := has_mod.mk bmod

definition bv_has_modular_arith [instance] (n : ℕ) : has_modular_arith (bv n) (nat_pow 2 n) :=
  has_modular_arith.mk (b_eq_modulo n)
