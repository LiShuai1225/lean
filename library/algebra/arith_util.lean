import algebra.ordered_field

definition div0 {A : Type} [field A] (x : A) : A := div x 0

constants (int rat real : Type.{1})
constants (int_locr : decidable_linear_ordered_comm_ring int)
constants (rat_lof : discrete_linear_ordered_field rat)
constants (real_lof : linear_ordered_field real)

attribute int_locr [instance]
attribute rat_lof [instance]
attribute real_lof [instance]

constants (real.of_rat : rat → real)
constants (rat.of_int : int → rat)
constants (int.of_nat : nat → int)

definition from_pos_num (A : Type) [has_zero A] [has_one A] [has_add A] : pos_num → A
| (pos_num.one) := 1
| (pos_num.bit0 k) := bit0 (from_pos_num k)
| (pos_num.bit1 k) := bit1 (from_pos_num k)

structure cyclic_numerals [class] (A : Type) extends comm_ring A :=
(bound : pos_num)  (cyclic : ∀ a k : A , a = a - k * from_pos_num A bound)

constant (bv2 : Type.{1})
constant (bv2_cr : field bv2)
attribute bv2_cr [instance]
definition bv2_cn [instance] : cyclic_numerals bv2 := ⦃ cyclic_numerals bv2, bv2_cr, bound := 4, cyclic := sorry ⦄

print cyclic_numerals.bound
