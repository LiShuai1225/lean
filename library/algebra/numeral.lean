import algebra.ring -- has_add, has_one, ... will be moved to init in the future
open algebra

variable {A : Type}

-- variables [s : ring A]
-- set_option pp.all true
-- check bit1 (bit0 (one : A))
