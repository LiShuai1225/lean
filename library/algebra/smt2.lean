import algebra.ordered_field

-- Core
constants (xor : Prop → Prop → Prop)
--constants (distinct: Π {A : Type}, A → A → Prop) -- not actually used

-- Arithmetic
constants (int real : Type.{1})

constants (int_has_add : has_add int)
constants (int_has_mul : has_mul int)
constants (int_has_sub : has_sub int)
constants (int_has_neg : has_neg int)
constants (int_has_div : has_div int)
constants (int_has_lt : has_lt int)
constants (int_has_le : has_le int)
constants (int_has_mod : has_mod int)
constants (int_decidable_linear_ordered_comm_group : decidable_linear_ordered_comm_group int)
attribute [instance] int_has_add int_has_mul int_has_sub int_has_div int_has_neg int_has_le int_has_lt int_has_mod int_decidable_linear_ordered_comm_group

constants (real_has_add : has_add real)
constants (real_has_mul : has_mul real)
constants (real_has_sub : has_sub real)
constants (real_has_neg : has_neg real)
constants (real_has_div : has_div real)
constants (real_has_lt : has_lt real)
constants (real_has_le : has_le real)

attribute [instance] real_has_add real_has_mul real_has_sub real_has_neg real_has_div real_has_le real_has_lt

constants (real.of_int : int → real) (real.to_int : real → int) (real.is_int : real → Prop)

-- Arrays
-- Note: we don't use map yet because of the [option] in the return type
-- Need: inhabited instances
constant (array : Type → Type → Type)
namespace array
constant (select : Π (A B : Type), A → array A B → B)

universe variables l₁ l₂ l₃
constant (store : Π (A : Type.{l₁}) (B : Type.{l₂}), A → B → array.{l₁ l₂ l₃} A B → array.{l₁ l₂ l₃} A B)
end array

namespace tactic
meta_definition smt2 : tactic unit := trace_state >> now
end tactic
