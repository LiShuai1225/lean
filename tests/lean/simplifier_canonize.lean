import algebra.group
open tactic

set_option simplify.theory false
set_option pp.implicit true

universe l

constants (f₁ : Π (X : Type) (X_grp : group X), X)
constants (f₂ : Π (X : Type) [X_grp : group X], X)
constants (A : Type.{l}) (A_grp₁ : group A)

definition A_grp₂ [irreducible] : group A := A_grp₁
definition A_grp₃ [irreducible] (t : true) : group A := A_grp₁

set_option simplify.canonize_fixed_point false
print "This should fail, since A_grp₂ is not unfolded, and it is not an instance arg"
example : @f₁ A A_grp₁ = @f₁ A A_grp₂ := by simp >> reflexivity

print "This should succeed, since A_grp₂ is an instance arg and gets canonized"
example : @f₂ A A_grp₁ = @f₂ A A_grp₂ := by simp >> reflexivity

print "This should fail to simplify, since we do not reach a fixed point"
example : @f₂ A (A_grp₃ trivial) = @f₂ A A_grp₂ := by simp

print "This should simplify, since we reach a fixed point"
set_option simplify.canonize_fixed_point true
example : @f₂ A (A_grp₃ trivial) = @f₂ A A_grp₂ := by simp
