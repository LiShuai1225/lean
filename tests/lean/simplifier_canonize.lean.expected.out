simplifier_canonize.lean:2:0: warning: imported file uses 'sorry'
simplifier_canonize.lean:13:11:failed to generate bytecode for 'A_grp₂'
code generation failed, VM does not have code for 'A_grp₁'
simplifier_canonize.lean:14:11:failed to generate bytecode for 'A_grp₃'
code generation failed, VM does not have code for 'A_grp₁'
This should fail, since A_grp₂ is not unfolded, and it is not an instance arg
simplifier_canonize.lean:18:41: error: simp tactic failed to simplify
state:
⊢ f₁ A A_grp₁ = f₁ A A_grp₂
simplifier_canonize.lean:18:41: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
This should succeed, since A_grp₂ is an instance arg and gets canonized
simplifier_canonize.lean:21:41: error: exact tactic failed, type mismatch, given expression has type
  @f₂ A A_grp₁ = @f₂ A A_grp₁
but is expected to have type
  @f₂ A A_grp₁ = @f₂ A A_grp₂
state:
Htarget : @f₂ A A_grp₁ = @f₂ A A_grp₁
⊢ @f₂ A A_grp₁ = @f₂ A A_grp₂

⊢ @f₂ A A_grp₁ = @f₂ A A_grp₁
simplifier_canonize.lean:21:41: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
This should fail to simplify, since we do not reach a fixed point
simplifier_canonize.lean:24:51: error: exact tactic failed, type mismatch, given expression has type
  @f₂ A A_grp₁ = @f₂ A A_grp₁
but is expected to have type
  @f₂ A (A_grp₃ trivial) = @f₂ A A_grp₂
state:
Htarget : @f₂ A A_grp₁ = @f₂ A A_grp₁
⊢ @f₂ A (A_grp₃ trivial) = @f₂ A A_grp₂

⊢ @f₂ A A_grp₁ = @f₂ A A_grp₁
simplifier_canonize.lean:24:51: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
This should simplify, since we reach a fixed point
simplifier_canonize.lean:28:51: error: exact tactic failed, type mismatch, given expression has type
  @f₂ A A_grp₂ = @f₂ A A_grp₂
but is expected to have type
  @f₂ A (A_grp₃ trivial) = @f₂ A A_grp₂
state:
Htarget : @f₂ A A_grp₂ = @f₂ A A_grp₂
⊢ @f₂ A (A_grp₃ trivial) = @f₂ A A_grp₂

⊢ @f₂ A A_grp₂ = @f₂ A A_grp₂
simplifier_canonize.lean:28:51: error: failed to add declaration 'example' to environment, value has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  ?M_1
